src/marks/text.js

```diff
@@ -375,23 +396,126 @@ const defaultWidthMap = {
+    j = readCharacter(text, i); // read the next character
+    const l = widthof(text, i, j); // current character width
+    if (w + l > width) {
+      w += inset;
+      while (w > width && i > 0) (j = i), (i = I.pop()), (w -= widthof(text, i, j)); // remove excess
+      return [i, width - w];
+    }
+    w += l;
+    I.push(i);
+  }
+  return [-1, 0];
+}
+
+export function clipEnd(text, width, widthof, ellipsis) {
+  text = text.trim(); // ignore leading and trailing whitespace
+  const e = widthof(ellipsis);
+  const [i] = cut(text, width, widthof, e);
+  return i < 0 ? text : text.slice(0, i).trimEnd() + ellipsis;
+}
+
+export function clipMiddle(text, width, widthof, ellipsis) {
+  text = text.trim(); // ignore leading and trailing whitespace
+  const w = widthof(text);
+  if (w <= width) return text;
+  const e = widthof(ellipsis) / 2;
+  const [i, ei] = cut(text, width / 2, widthof, e);
+  const [j] = cut(text, w - width / 2 - ei + e, widthof, -e); // TODO read spaces?
+  return j < 0 ? ellipsis : text.slice(0, i).trimEnd() + ellipsis + text.slice(readCharacter(text, j)).trimStart();
+}
+
+export function clipStart(text, width, widthof, ellipsis) {
+  text = text.trim(); // ignore leading and trailing whitespace
+  const w = widthof(text);
+  if (w <= width) return text;
+  const e = widthof(ellipsis);
+  const [j] = cut(text, w - width + e, widthof, -e); // TODO read spaces?
+  return j < 0 ? ellipsis : ellipsis + text.slice(readCharacter(text, j)).trimStart();
+}
+
+// TODO I wrote these as regular expressions for clarity, but we might want to
+// optimize by testing char codes directly for surrogate pair and zero-width
+// joiner detection (assuming that doing so would actually be faster in a
+// material way… maybe a quick microbenchmark to see if it matters)?
+const reSurrogatePair = /[\uD800-\uDBFF][\uDC00-\uDFFF]/y;
+const reCombiner = /[\p{Combining_Mark}\p{Emoji_Modifier}]+/uy;
+const reZeroWidthJoiner = /\u200D/y;
+const rePictographic = /\p{Extended_Pictographic}/uy;
+
+// Reads a single “character” element from the given text starting at the given
+// index, returning the index after the read character. Ideally, this implements
+// the Unicode text segmentation algorithm and understands grapheme cluster
+// boundaries, etc., but in practice this is only smart enough to detect UTF-16
+// surrogate pairs, combining marks, and zero-width joiner (zwj) sequences such
+// as emoji skin color modifiers. https://unicode.org/reports/tr29/
+export function readCharacter(text, i) {
+  reSurrogatePair.lastIndex = i++; // consume first character
+  if (reSurrogatePair.test(text)) ++i; // consume surrogate pair
+  reCombiner.lastIndex = i;
+  if (reCombiner.test(text)) i = reCombiner.lastIndex; // consume combiners
+  reZeroWidthJoiner.lastIndex = i;
+  if (reZeroWidthJoiner.test(text)) return readCharacter(text, i + 1); // consume zero-width join and what follows
+  return i;
+}
```

Comment:
```suggestion
const reSurrogatePair = /[\uD800-\uDBFF][\uDC00-\uDFFF]/y;
const reCombiner = /[\p{Combining_Mark}\p{Emoji_Modifier}]+/uy;
const rePictographic = /\p{Extended_Pictographic}/uy;

// Reads a single “character” element from the given text starting at the given
// index, returning the index after the read character. Ideally, this implements
// the Unicode text segmentation algorithm and understands grapheme cluster
// boundaries, etc., but in practice this is only smart enough to detect UTF-16
// surrogate pairs, combining marks, and zero-width joiner (zwj) sequences such
// as emoji skin color modifiers. https://unicode.org/reports/tr29/
export function readCharacter(text, i) {
  i++; // consume first character
  if (text.charCodeAt(i) >= 0xd800 && ((reSurrogatePair.lastIndex = i - 1), reSurrogatePair.test(text))) ++i; // consume surrogate pair
  if (text.charCodeAt(i) > 0x7f) {
    reCombiner.lastIndex = i;
    if (reCombiner.test(text)) i = reCombiner.lastIndex; // consume combiners
  }
  if (text.charCodeAt(i) === 0x200d) return readCharacter(text, i + 1); // consume zero-width join and what follows
  return i;
}
```

this seems quite a bit faster on the three browsers I've tested:
https://observablehq.com/@observablehq/benchmark-readchar-1283

Reply:
Thanks. I want to try a few additional tweaks using charCodeAt and I’ll push a commit if it works. (Not a blocker though; I expect we could do this performance improvement as a followup.)

Reply:
Tidied up in 954ee22.